### PR TITLE
DRA: add default claim parameters to resource class

### LINKED ISSUE #123858: DRA: structured parameters: handling of claim without claim parameters
### What would you like to be added?

Without claim parameters, it is unclear which structured model is meant to be used. If even it was clear, parameters for it might be useful.

We should add a default parameter reference to the ResourceClass. Then if a ResourceClaim has no parameter reference, that default gets used. If both are unset, the claim cannot get allocated.

### Why is this needed?

A claim without parameters may make sense depending on the hardware when users have no special needs.

### PR DESCRIPTION
**Pull request by ravisantoshgudimetla**
#### What type of PR is this?

/kind feature
#### What this PR does / why we need it:
Add a default parameter reference to the ResourceClass. This should be created by cluster admin while creating resource class.
#### Which issue(s) this PR fixes:

Fixes #123858

#### Does this PR introduce a user-facing change?

```release-note
Add a default parameter reference to the ResourceClass. This should be created by cluster admin while creating resource class.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by ravisantoshgudimetla** [comment]
cc @pohly

File: pkg/apis/resource/types.go [hunk 1]
```diff
@@ -366,6 +366,15 @@ type ResourceClass struct {
 	// If and only if allocation of claims using this class is handled
 	// via structured parameters, then StructuredParameters must be set to true.
 	StructuredParameters *bool
+
+	// DefaultClaimParametersRef is an optional reference to an object that holds parameters
+	// used by the driver as default when allocating a resource associated with this class.
+	// It is the responsibility of the cluster administrator to create this default resource
+	// claim parameter reference when defining the resource class. This field is utilized
+	// when the ParametersRef of the resource class is nil. If both ParametersRef
```
Comment (pohly) [inline pkg/apis/resource/types.go:374]:
```suggestion
	// only when the ParametersRef of the resource claim is nil. If both ParametersRef
```

**Review by pohly — COMMENTED** [review]
Can you include the code which uses the new field?

**Comment by ravisantoshgudimetla** [comment]
> Can you include the code which uses the new field?

You mean the scheduler changes? I'd typically put them in a separate PR to make sure that api folks can review it easily.

**Comment by pohly** [comment]
Yes. Please add here in a separate commit. We don't want to end up in a situation where an API change is merged, but then the implementation doesn't work as intended or gets delayed.

File: staging/src/k8s.io/kms/internal/plugins/_mock/go.work.sum [hunk 2]
```diff
@@ -24,6 +24,8 @@ github.com/stretchr/objx v0.1.0 h1:4G4v2dO3VZwixGIRoQ5Lfboy6nUhCyYzaqnIAPPhYs4=
 github.com/yuin/goldmark v1.2.1 h1:ruQGxdhGHe7FWOJPT0mKs5+pD2Xs1Bm/kdGlHO04FmM=
 golang.org/x/crypto v0.19.0 h1:ENy+Az/9Y1vSrlrvBSyna3PITt4tiZLf7sgCjZBX7Wo=
 golang.org/x/crypto v0.19.0/go.mod h1:Iy9bg/ha4yyC70EfRS8jz+B6ybOBKMaSxLj6P6oBDfU=
+golang.org/x/crypto v0.21.0 h1:X31++rzVUdKhX5sWmSOFZxx8UW/ldWx55cbf08iNAMA=
```
Comment (aramase) [inline staging/src/k8s.io/kms/internal/plugins/_mock/go.work.sum:27]: Is this change required? Could you revert this?

Comment (ravisantoshgudimetla) [reply staging/src/k8s.io/kms/internal/plugins/_mock/go.work.sum]: This is not required. This is coming from me running `make update` which updates the workspaces. And btw, if anyone runs it, they might still get it. I've removed it for now.

File: pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go [hunk 3]
```diff
@@ -1016,30 +1016,24 @@ func (pl *dynamicResources) lookupClassParameters(logger klog.Logger, class *res
...
-							NamedResources: &resourcev1alpha2.NamedResourcesRequest{
-								Selector: "true",
-							},
-						},
-					},
-				},
-			},
-		},
-	}
+	if claim.Spec.ParametersRef == nil && class.DefaultClaimParametersRef == nil {
```
Comment (ravisantoshgudimetla) [inline pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:1019]: @pohly - Are these 2 the only cases that we should be concerned about? Let me know if I'm missing something.

File: pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go [hunk 4]
```diff
@@ -1016,30 +1016,24 @@ func (pl *dynamicResources) lookupClassParameters(logger klog.Logger, class *res
...
+		return nil, statusError(logger, fmt.Errorf("error allocating claim, either claim's parameter reference or default parameters for resourceclass need to be set"))
```
Comment (ravisantoshgudimetla) [inline pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:1020]: and also should we return statusError or statusUnScheduleable?

Comment (pohly) [reply pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:1020]: `statusUnscheduleable` - the pod is permanently unschedulable until claim or class are changed or the parameters are created. The plugin reacts to cluster events for that, so there is no need for periodic retries, which is what `statusError` implies.

**Comment by pohly** [comment]
/retitle DRA: add default claim parameters to resource class

**Comment by pohly** [comment]
There's a known data race in the dynamicresources unit test, fix is in https://github.com/kubernetes/kubernetes/pull/124409

**Comment by pohly** [comment]
@ravisantoshgudimetla: Ignore the retests... the unit test failure is genuine and seems related to changes in this PR.

**Comment by pohly** [comment]
/hold

We need to decide what the desired semantic is for this new field. See https://github.com/kubernetes/enhancements/pull/4585#discussion_r1574628189.

Once that KEP update is merged, we can update this PR accordingly (API wording, functionality).

**Comment by ravisantoshgudimetla** [comment]
@pohly - Just catching up on this after coming from break, so the consensus seems to be https://github.com/kubernetes/enhancements/pull/4585#discussion_r1582638831, which means, there will be no allocation when neither claim not class exists, so this PR can continue?

**Comment by pohly** [comment]
There are more API changes coming: https://github.com/kubernetes-sigs/wg-device-management/pull/14

We might not have a ResourceClass that gets referenced by a ResourceClaim - let's wait for that decision.

**Comment by pohly** [comment]
The new API in 1.31 no longer needs this. Nonetheless, thanks for your help! I'm sure we can find other things to do in this space, join us in #wg-device-management.

/close
